tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/predicates.md

```diff
@@ -0,0 +1,352 @@
+
+return state.db.all(sql, ...parameters);
+```
+
+Let's delegate to a helper function in order to build our `WHERE` clause. Let's call our function `visit_expression`,
+because we're using the visitor design pattern.
+
+```typescript
+const where_clause = request.query.where == null ? "" : `WHERE ${visit_expression(parameters, request.query.where)}`;
+```
+
+We'll handle each different type of expression by pattern matching on the `type` field of the current expression.
+
+```typescript
+function visit_expression(parameters: any[], expr: Expression): string {
+  switch (expr.type) {
+    case "and":
+
+    case "or":
+
+    case "not":
+
+    case "unary_comparison_operator":
+
+    case "binary_comparison_operator":
+
+    case "binary_array_comparison_operator":
+
+    case "exists":
+
+    default:
+      throw new BadRequest("Unknown expression type");
+  }
+}
+```
+
+For the logical expressions `and` and `or`, we will visit each of the subexpressions in turn, and concatenate the
+generated SQL.
+
+```typescript
+// ...
+case "and":
+    if (expr.expressions.length > 0) {
+        return expr.expressions.map(e => visit_expression_with_parens(parameters, e)).join(" AND ");
+    } else {
+        return "TRUE";
+    }
+// ...
+```
+
+We need a helper function here, which visits an expression, but always wraps the results in parentheses. This way, we
+don't generate SQL with the wrong operator precedence in cases where logical operators are nested.
+
+```typescript
+function visit_expression_with_parens(parameters: any[], expr: Expression): string {
+    return `(${visit_expression(parameters, expr)})`;
+}
+```
+
+Note that we also need to make a special case for zero subexpressions, or else we'll generate invalid SQL.
+
+We can complete the `or` and `not` cases very similarly.
+
+```typescript
+// ...
+case "or":
+    if (expr.expressions.length > 0) {
+        return expr.expressions.map(e => visit_expression_with_parens(parameters, e)).join(" OR ");
+    } else {
+        return "FALSE";
+    }
+case "not":
+    return `NOT ${visit_expression_with_parens(parameters, expr.expression)}`;
+// ...
+```
+
+For `unary_comparison_operator` expressions, we can switch on `expr.operator`. Right now, the only option is the
+`is_null` operator:
+
+```typescript
+// ...
+case "unary_comparison_operator":
+    switch (expr.operator) {
+      case 'is_null':
+        return `${visit_comparison_target(expr.column)} IS NULL`;
+      default:
+        throw new BadRequest("Unknown comparison operator");
+    }
+// ...
+```
+
+For `binary_comparison_operator` expressions, we can switch on `expr.operator.type`.  We will only implement the `equal`
+operator, because our schema doesn't advertise any custom binary operators. If we wanted to add another operator, like a
+"greater than" operator for numbers, we would do that here, and also advertise that operator in the NDC schema response.
+
+```typescript
+// ...
+switch (expr.operator.type) {
```

Comment:
It would help to have the `binary_comparison_operator` switch statement included in this snippet like you did for `unary_comparison_operator`.

Reply:
👍